translation_layer/jpl-capi.js

```diff
@@ -0,0 +1,150 @@
+    var Types = require('api/snapshot/SimCapiValue').TYPES;
+
+	var initialized = false;
+
+    window.receiveExposeFromRanger = function (name, type, value, allowedValues)
+    {
+        if (!initialized)
+        {
+            Controller.notifyOnReady();
+            capi = new Capi();
+            capi.on('change', onCapiChange);
+            initialized = true;
+        }
+
+        var typeConverter = generateTypeConverter(type);
+        value = typeConverter(value);
+
+        var descObj = {};
+
+        //console.log("allowed values: " + allowedValues);
+
+        if (allowedValues)
+        {
+            descObj.allowedValues = allowedValues.map(typeConverter);
+        }
+
+        capi.set(name, value);
+        adapter.expose(name, capi, descObj);
+    };
+
+	    window.receiveValueFromRanger = function (name, type, value) {
+        if (initialized) {
+            value = generateTypeConverter(type)(value);
+            capi.set(name, value);
+        }
+    };
+
+    window.UpdateScreenAtSPR = function (_func) 
+    {
+        Transporter.triggerCheck();
+    }
+	
+    function storageSuccess() {
+        console.log("Storage process successful for ");
+        console.log(Transporter.getHandshake().authToken);
+    };
+
+    function storageError() {
+        console.log("Storage process errored!");
+    };
+	
+	function transmitDataToSim(value)
+    {
+        console.log(value.value.toString());
+        // do a thing, preferrably where you take data from SPR and give it to Ranger
+    }
+
+    // Utilize the transporter for more permanent storage.
+    // JOS: 12/9/2016
+    window.storeRangerData = function (simId, key, value)
+    {
+        Transporter.setDataRequest(simId, key, value, storageSuccess, storageError);
+    };
+
+    window.getRangerData = function (simId, key)
+    {
+        Transporter.getDataRequest(simId, key, transmitDataToSim, storageError);
+    };
+
+    // TODO pregenerate for each type and cache
+    function generateTypeConverter (type) {
+        if (type === Types.NUMBER) {
+            return function (value) {
+                return parseFloat(value);
+            };
+        } else if (type === Types.BOOLEAN) {
+            return function (value) {
+                return value === "True" ? true : false;
+            };
+        }
+
+        // for strings and arrays do nothing
+        return function (value) { return value; };
+    }
+
+    function onCapiChange () {
+        var changedAttributes = capi.changedAttributes();
+        if (changedAttributes) {
+            for (var i in changedAttributes)
+            {
+                console.log("setting " + i + " to " + changedAttributes[i]);
+				
+				if(typeof ranger_eclipse !== 'undefined')
+				{
+					//de nada
+				}
+				else
+				{
+					console.log("Ranger not found");
+					return;
+				}
+				
+				switch(i)
+				{
+					case "System.Show Shadow Labels":
```

Comment:
I think I'd probably create a map for these simple fields like:
```
var aliasMap = {
  "System.Show Shadow Labels": "shadowLabelsEnabled",
  "System.Camera FOV": "setFOV"
}
```
which would allow you to simplify this logic (and hopefully remove the switch) to something like:
```
var rangerMessage = {};
if(aliasMap.hasOwnProperty(i)) {
  rangerMessage[aliasMap[i]] = changedAttributes[i];
} else { 
// ... other messages like Add Location
}
ranger_eclipse.update(rangerMessage);
```